src/plugin/ipc/file/fileconnection.cpp

```diff
@@ -821,6 +829,20 @@ void FileConnection::postRestart()
 
     if (fd == -1) {
       _fileAlreadyExists = true;
+      bool overwrite_file = false;
+      if(dmtcp_overwrite_ckptfile_on_restart &&
+          dmtcp_overwrite_ckptfile_on_restart(_path.c_str())) {
+        overwrite_file = true;
+      }
+      if(overwrite_file) {
```

Comment:
This is probably not the best place to overwrite a file with its ckpt-ed copy; you need to defer it to the `refill()` method. See the comment above the `_real_open()` call.

The only guarantee provided in a distributed system with barriers is that a process will wait until all its peers reach the same barrier, but it is free to process its callbacks for the barrier independent of the others. And so, one reason for the open() call returning a -1 is simply that some other restarting process succeeded in creating the file before the current process got to this point, and hence, it's possible that there's data in flight.